bentoml/_internal/io_descriptors/numpy.py

```diff
@@ -546,15 +359,176 @@ async def predict(input: NDArray[np.int16]) -> NDArray[Any]:
+            )
+            shape = self._shape
+        else:
+            shapepb = field.shape  # this can be an empty list ([])
+            if shapepb:
+                shape = tuple(shapepb)
+                if self._shape:
+                    logger.warning(
+                        f"'shape={self._shape}' is ignored when 'shape' is specified in request message. Using '{shape}' from request message..."
+                    )
+            else:
+                shape = self._shape
+
+        # The same logic for 'shape' applies to 'dtype' with the additional:
+        #
+        # 2a. If 'dtype' message is not specified, then we will infer given numpy dtype from the entry.
+        if self._enforce_dtype:
+            logger.debug(
+                f"'enforce_dtype={self._enforce_dtype}', using '{self._dtype}'..."
+            )
+            dtype = self._dtype
+            field_values = npdtype_to_fieldpb_map()[dtype]
+
+            values_arr = getattr(field, field_values, None)
+            if not values_arr:
+                raise_grpc_exception(
+                    f"field '{field_values}' (required for '{dtype}') is missing.",
+                    context=context,
+                    exception_cls=UnprocessableEntity,
+                )
+        else:
+            dtypepb = field.dtype
+            if dtypepb:
+                if dtypepb == pb.NDArray.DTYPE_UNSPECIFIED:
+                    dtype = None
+                else:
+                    dtype = dtypepb_to_npdtype_map()[dtypepb]
+
+                if self._dtype:
+                    logger.warning(
+                        f"'dtype={self._dtype}' is ignored when 'dtype' is specified in request message. Using '{dtype}' from request message..."
+                    )
+
+                values_arr = getattr(field, dtypepb_to_fieldpb_map()[dtypepb])
+
+                if not dtype:
+                    # This is the case where 'dtype' proto uses DTYPE_UNSPECIFIED
+                    return np.empty(shape=shape or 0)
+            else:
+                fieldpb = [
+                    f.name for f, _ in field.ListFields() if f.name.endswith("_values")
+                ]
+                if len(fieldpb) != 1:
+                    raise_grpc_exception(
+                        f"Array contents can only be one of given values key. Use one of {fieldpb} instead.",
+                        context=context,
+                        exception_cls=BadInput,
+                    )
+                dtype = fieldpb_to_npdtype_map()[fieldpb[0]]
+                values_arr = getattr(field, fieldpb[0])
+
+        if shape and dtype in [np.float32, np.double, np.bool_]:
+            buffer = field.SerializeToString()
+            num_entries = np.prod(shape)
+            return np.frombuffer(
+                memoryview(buffer[-(dtype.itemsize * num_entries) :]),
+                dtype=dtype,
+                count=num_entries,
+                offset=0,
+            ).reshape(shape)
```

Comment:
add self.validate_array here

Reply:
I think then we don't need to `reshape` here.

Reply:
No need to right? we already cast the dtype and reshape it with `shape`